content/entry-hash/index.md

```diff
@@ -0,0 +1,219 @@
+---
+rfc:
+start_date: 2018-08-02
+pr:
+status: draft
+---
+
+# Entry hash
+
+## Summary
+
+This RFC proposes an algorithm to hash entries with a well defined set of
+attributes, and a way to hash from the values to the whole entry.
+
+**This RFC is incompatible with the reference implementation**
+
+## Motivation
+
+The current specfication doesn't define what is the algorithm to get from the
+entry data to the entry hash for the given hashing algorithm (from now on
+SHA-256).
+
+The reference implementation (ORJ) reuses the item canonicalisation
+algorithm to canonicalise the entry JSON representation and then hash the
+result. This means that any attribute currently part of the reference
+implementation `Entry` entity is serialised and used to compute the hash. As
+an example, the index entry number, part of the experimental work on Indexes
+has affected the result of the entry hash even though it is (a) not stable and
+(b) not part of the real identity of an entry.
+
+Finally, the current implementation relies on JSON and on the attribute names
+serialised as strings.
+
+This RFC aims to provide an algorithm that doesn't rely on specific attribute
+names (e.g. `entry-number` vs `number`, `item-hash` vs `itemHash`, etc) which
+supports a better decoupling between the data model and the REST API (we don't
+have plans for it but a hypothetical GraphQL interface wouldn't work with the
+current attribute names and yet conceptually an entry is the same).
+
+Note that the item hash algorithm is not affected by this RFC.
+
+## Explanation
+
+This section describes the algorithm twice. First in a high level and then a
+more step by step with intermediate values.
+
+For reference, this is the abstract definition of an entry:
+
+```elm
+type Entry =
+  { number : Integer
+  , key : ID
+  , timestamp : Timestamp
+  , item : Set Hash
+  }
+```
+
+### The algorithm
+
+When this algorithm operates on _hashes_ (e.g. tag, concatenate) it is done on
+the byte level, not the hexadecimal string representation that the latter
+example shows as partial representations.
+
+1. Let _number_ be the string representation of the entry number.
+2. Let _key_ be the string representation of the entry key.
+3. Let _timestamp_ be the string representation of the entry timestamp.
+4. Let _items_ be the set of item hashes as bytes. Note that
+   the item hashes must not have the prefix (i.e. `sha-256:`) when transformed to
+   bytes.
```

Comment:
Mostly because the hash is treated as a byte sequence and the `sha-256` is not part of the hash. When treated like a byte sequence, you don't need to hash it again (raw), you can use it as is, which is similar to how redaction works.

I've always seen the hashes with the prefix as a string representation of a qualified sequence of bytes encoded in hexadecimal. That's probably the reason.